common-pg/src/main/scala/com/socrata/pg/query/QueryServerHelper.scala

```diff
@@ -105,23 +107,44 @@ object QueryServerHelper {
     ) with RowReaderQuerier[SoQLType, SoQLValue]
   }
 
-  def getJoinCopies(pgu: PGSecondaryUniverse[SoQLType, SoQLValue],
-                    analyses: BinaryTree[SoQLAnalysis[UserColumnId, SoQLType]],
-                    reqCopy: Option[String]): Map[TableName, CopyInfo] = {
+  def collectRelatedTableNames(analyses: BinaryTree[SoQLAnalysis[UserColumnId, SoQLType]]): Seq[TableName] = {
     val joins = typed.Join.expandJoins(analyses.seq)
     val froms = analyses.seq.flatMap(x => x.from.toSeq)
-    val joinTables = joins.flatMap(x => x.from.fromTables) ++ froms
-    val joinTableMap = joinTables.flatMap { resourceName =>
-      getCopy(pgu, new ResourceName(resourceName.name), reqCopy).map(copy => (resourceName, copy))
-    }.toMap
-
-    val allMap = joinTableMap
-    val allMapNoAliases = allMap.map {
-      case (tableName, copy) =>
-        (tableName.copy(alias = None), copy)
-    }
+    joins.flatMap(x => x.from.fromTables) ++ froms
+  }
 
-    allMapNoAliases
+  def removeTableAlias(tableNames: Seq[TableName]): Seq[TableName] = {
+    tableNames.map(_.copy(alias = None))
+  }
+
+  def getCopyAndRollupMaps(pgu: PGSecondaryUniverse[SoQLType, SoQLValue],
+                           tableNames: Seq[TableName],
+                           reqCopy: Option[String]):
+      (Map[TableName, CopyInfo], Map[TableName, LocalRollupInfo]) = {
+    val rollupRx = "(_.{4}-.{4})\\.(.+)".r
+    tableNames.foldLeft((Map.empty[TableName, CopyInfo], Map.empty[TableName, LocalRollupInfo])) { (acc, tableName) =>
+      tableName.name match {
+        case rollupRx(resourceName, rollupName) =>
+          val ruOpt = for {
+            datasetInfo <- pgu.datasetMapReader.datasetInfoByResourceName(ResourceName(resourceName))
+            copyInfo <- pgu.datasetMapReader.lookup(datasetInfo, LifecycleStage.Published)
+            rollupInfo <- pgu.datasetMapReader.rollup(copyInfo, new RollupName(rollupName))
+          } yield { rollupInfo }
```

Comment:
Side comment: When do you decide to use for/yield instead of map()?

Reply:
A for comprehension is multiple flatmaps and map.